### PR TITLE
[5398] - undo withdrawal

### PR DESCRIPTION
**Pull request by d-a-v-e**
### Context

We want to let admins undo withdrawals using our UI.

Our current support playbook suggests these actions should happen without audits - we think it’s better that all actions be audited - including collecting a comment and zendesk url.

[Link to PR adding this to the prototype](https://github.com/DFE-Digital/register-trainee-teachers-prototype/pull/626)

### Changes proposed in this pull request

* adds new undo withdrawal journey
* adds new timeline element for undo withdrawal audit.

### Guidance

https://user-images.githubusercontent.com/6339025/233941771-2feb0b02-f366-4b14-8d1a-d69a421c629d.mov

### CODE REVIEW

**Comment by d-a-v-e** [comment]
[[L] Add feature to allow admins to undo a withdrawal](https://trello.com/c/6PrddQm5/5398-l-add-feature-to-allow-admins-to-undo-a-withdrawal)

File: app/forms/undo_withdrawal_form.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+class UndoWithdrawalForm
+  include ActiveModel::Model
+
+  attr_accessor :comment, :ticket, :trainee, :session
+
+  validates :comment, presence: true
+  validates :ticket, format: URI::DEFAULT_PARSER.make_regexp(%w[http https]), allow_blank: true
+
+  def initialize(trainee: nil, comment: nil, ticket: nil, session: nil)
```
Comment (jhenderson) [inline app/forms/undo_withdrawal_form.rb:11]: It looks like `trainee` and `session` are minimal requirements.

```suggestion
  def initialize(trainee:, session:, comment: nil, ticket: nil)
```

File: app/services/trainees/create_timeline_events.rb [hunk 2]
```diff
@@ -121,6 +130,31 @@ def call
 
     attr_reader :audit, :current_user
 
+    def withdrawal_undone?
+      return unless audited_changes.is_a?(Hash)
+
+      audited_changes.any? do |key, value|
+        key == "state" && value.is_a?(Array) && value.size == 2 && value[0] == 4 && value[1].is_a?(Integer)
```
Comment (jhenderson) [inline app/services/trainees/create_timeline_events.rb:137]:
```suggestion
        key == "state" && value.is_a?(Array) && value.size == 2 && value[0] == Trainee.states["withdrawn"] && value[1].is_a?(Integer)
```

File: spec/features/trainee_actions/undo_withdraw_trainee_spec.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        fill_in "undo-withdrawal-form-ticket-field", with: "asdf"
+        click_button "Continue"
+        expect(page).to have_content("Must be a URL")
+      end
+    end
+  end
+
+  context "undo trainee withdrawal flow" do
+    scenario "successfully undo trainee withdrawal" do
+      fill_in "undo-withdrawal-form-comment-field", with: "this is the comment"
```
Comment (jhenderson) [inline spec/features/trainee_actions/undo_withdraw_trainee_spec.rb:38]: Can we wrap some of the low level steps into sentences? I know it's a bit of a pain, but we should try and be consistent with the other feature files.

**Review by jhenderson — APPROVED** [review]
Looks really good. Tested it out manually. Couldn't see any issues. Codewise, I've made some minor suggestions.

File: app/forms/undo_withdrawal_form.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    [comment, ticket].compact.join("\n")
+  end
+
+  def previous_state_enum
+    @previous_state_enum ||=
+      trainee
+        .audits
+        .where(action: :update)
+        .pluck(:audited_changes)
+        .reject { |hash| hash.dig("state", 1) == 4 }
```
Comment (jhenderson) [inline app/forms/undo_withdrawal_form.rb:54]:
```suggestion
        .reject { |hash| hash.dig("state", 1) == Trainee.states["withdrawn"] }
```

**Comment by edwardhorsford** [comment]
It looks like undoing a withdrawal and even editing is available to non admins.

Screenshot 2023-04-26 at 17 56 28

**Comment by edwardhorsford** [comment]
It looks like undoing a withdrawal and even editing is available to non admins.

Screenshot 2023-04-26 at 17 56 28

**Review by edwardhorsford — DISMISSED** [review]
It looks like undoing a withdrawal and even editing is available to non admins.

Screenshot 2023-04-26 at 17 56 28

**Comment by edwardhorsford** [comment]
Some other minor issues:

- Page 1: The page title is using title case rather than sentence case
- Page 2: The page title does not match the H1
- Page 2: Missing a caption with the trainee's name (screenshot below)
Screenshot 2023-04-26 at 18 03 41

- Page 4: I don't think we should be showing the url on the timeline:
Screenshot 2023-04-26 at 18 08 36

**Comment by d-a-v-e** [comment]
> It looks like undoing a withdrawal and even editing is available to non admins.
> 
> Screenshot 2023-04-26 at 17 56 28

The undo withdraw action is forbidden, but the button (and 1st page) were both appearing. I've fixed this now!

**Comment by edwardhorsford** [comment]
I don't think we should validate this field.
Screenshot 2023-04-27 at 14 16 39

**Comment by edwardhorsford** [comment]
The confirm page title does not match the h1.

**Pull request merged**